File: src/states/menu_state.py
"""Main menu – horizontal three-button layout with full mouse support."""
import math
import pygame
from src.states.base_state import BaseState
from src.config import (
    SCREEN_WIDTH, SCREEN_HEIGHT,
    UI_BG, UI_PANEL, UI_BORDER, UI_TEXT, UI_HEADING, UI_SUBTEXT,
    UI_HIGHLIGHT, UI_SUCCESS, ASTAR_MAN_COL, ASTAR_EUC_COL, DIJKSTRA_COL,
    font,
)

# ── Button layout constants ───────────────────────────────────────────────────
_BTN_W   = 210
_BTN_H   = 240
_BTN_GAP = 50
_BTN_TOP = (SCREEN_HEIGHT - _BTN_H) // 2 - 20   # vertical centre with offset for title
_TOTAL_W = 3 * _BTN_W + 2 * _BTN_GAP
_BTN_X0  = (SCREEN_WIDTH - _TOTAL_W) // 2

_BUTTONS = [
    {"label": "Tutorial", "sublabel": "Learn A* step by step", "state": "TUTORIAL", "key": pygame.K_t},
    {"label": "Maze",     "sublabel": "Race through 4 layers", "state": "MAZE_L1",  "key": pygame.K_1},
    {"label": "Custom",   "sublabel": "Coming soon…",          "state": "CUSTOM",   "key": pygame.K_c},
]


class MenuState(BaseState):

    def enter(self) -> None:
        self._hover   = -1   # index of hovered button (-1 = none)
        self._anim    = 0.0
        # Build rects for hit-testing
        self._rects = [
            pygame.Rect(_BTN_X0 + i * (_BTN_W + _BTN_GAP), _BTN_TOP, _BTN_W, _BTN_H)
            for i in range(len(_BUTTONS))
        ]

    # ── Events ────────────────────────────────────────────────────────────────

    def handle_event(self, event: pygame.event.Event) -> None:
        if event.type == pygame.MOUSEMOTION:
            self._hover = -1
            for i, r in enumerate(self._rects):
                if r.collidepoint(event.pos):
                    self._hover = i
                    break

        elif event.type == pygame.MOUSEBUTTONDOWN and event.button == 1:
            for i, r in enumerate(self._rects):
                if r.collidepoint(event.pos):
                    self._go(_BUTTONS[i]["state"])
                    return

        elif event.type == pygame.KEYDOWN:
            for btn in _BUTTONS:
                if event.key == btn["key"]:
                    self._go(btn["state"])
                    return
            # ESC / Q quits to OS level is intentionally not handled here.

    # ── Update ────────────────────────────────────────────────────────────────

    def update(self, dt: float) -> None:
        self._anim += dt

    # ── Draw ──────────────────────────────────────────────────────────────────

    def draw(self) -> None:
        s = self.screen
        s.fill(UI_BG)

        # ── Animated background grid dots ─────────────────────────────────
        t = self._anim
        for gx in range(0, SCREEN_WIDTH, 60):
            for gy in range(0, SCREEN_HEIGHT, 60):
                alpha = int(30 + 20 * math.sin(t + gx * 0.05 + gy * 0.05))
                pygame.draw.circle(s, (alpha, alpha, alpha + 20), (gx, gy), 1)

        # ── Title ─────────────────────────────────────────────────────────
        cx = SCREEN_WIDTH // 2
        title_cn = font("cn_xl").render("算法之路", True, UI_HEADING)
        title_en = font("title").render("The Path of Algorithms", True, (180, 180, 200))
        sub      = font("body").render("A*  vs  Dijkstra  ·  Interactive Maze Learning", True, UI_SUBTEXT)
        s.blit(title_cn, (cx - title_cn.get_width() // 2, 60))
        s.blit(title_en, (cx - title_en.get_width() // 2, 106))
        s.blit(sub,      (cx - sub.get_width()     // 2, 140))

        # Decorative algorithm badges
        badges = [
            ("Dijkstra  h=0",            DIJKSTRA_COL),
            ("A* Manhattan  h=|Δx|+|Δy|", ASTAR_MAN_COL),
            ("A* Euclidean  h=√(Δx²+Δy²)", ASTAR_EUC_COL),
        ]
        bx = cx - 370
        for label, col in badges:
            bs = font("small").render(label, True, col)
            pygame.draw.rect(s, (*col, 40), (bx - 4, 168, bs.get_width() + 8, 20),
                             border_radius=3)
            s.blit(bs, (bx, 168))
            bx += bs.get_width() + 24

        # ── Three main buttons ─────────────────────────────────────────────
        for i, (btn, rect) in enumerate(zip(_BUTTONS, self._rects)):
            self._draw_button(s, i, btn, rect)

        # ── Key hints below buttons ────────────────────────────────────────
        hint_y = _BTN_TOP + _BTN_H + 20
        hint = font("small").render(
            "Click a button  ·  or press  [T] Tutorial   [1] Maze   [C] Custom",
            True, UI_SUBTEXT,
        )
        s.blit(hint, (cx - hint.get_width() // 2, hint_y))

        # ── Collected tomes ────────────────────────────────────────────────
        if self.game_data.get("tomes"):
            ty = SCREEN_HEIGHT - 38
            ts = font("small").render("Unlocked:", True, UI_HEADING)
            s.blit(ts, (30, ty))
            tx = 30 + ts.get_width() + 10
            for tome in self.game_data["tomes"]:
                col = ASTAR_MAN_COL if tome == "manhattan" else ASTAR_EUC_COL
                ls  = font("small").render(tome.capitalize() + " Tome ✦", True, col)
                s.blit(ls, (tx, ty))
                tx += ls.get_width() + 14

    # ── Single button draw ────────────────────────────────────────────────────

    def _draw_button(self, s, idx, btn, rect):
        hovered = (self._hover == idx)
        pulse   = 0.5 + 0.5 * math.sin(self._anim * 2 + idx * 1.2)

        # Background
        base_col = (45, 45, 68) if hovered else (30, 30, 48)
        pygame.draw.rect(s, base_col, rect, border_radius=16)

        # Glowing border on hover
        border_col = UI_HIGHLIGHT if hovered else UI_BORDER
        if hovered:
            # Outer glow
            glow_r = pygame.Rect(rect.x - 2, rect.y - 2, rect.w + 4, rect.h + 4)
            pygame.draw.rect(s, (*UI_HIGHLIGHT, int(60 + 40 * pulse)),
                             glow_r, width=2, border_radius=18)
        pygame.draw.rect(s, border_col, rect, width=2, border_radius=16)

        # Icon area (top 130 px of button)
        icon_cx = rect.centerx
        icon_cy = rect.y + 75

        if idx == 0:        # Tutorial – book icon
            _draw_book_icon(s, icon_cx, icon_cy, hovered)
        elif idx == 1:      # Maze – mini maze icon
            _draw_maze_icon(s, icon_cx, icon_cy, hovered)
        else:               # Custom – pencil + gear icon
            _draw_custom_icon(s, icon_cx, icon_cy, hovered)

        # Main label
        lbl_surf = font("heading").render(btn["label"], True,
                                          UI_HEADING if hovered else UI_TEXT)
        s.blit(lbl_surf, (icon_cx - lbl_surf.get_width() // 2, rect.y + 152))

        # Sublabel / description
        sub_surf = font("small").render(btn["sublabel"], True,
                                        UI_HIGHLIGHT if hovered else UI_SUBTEXT)
        s.blit(sub_surf, (icon_cx - sub_surf.get_width() // 2, rect.y + 184))

        # Key hint chip
        key_labels = {"TUTORIAL": "[T]", "MAZE_L1": "[1]", "CUSTOM": "[C]"}
        kl = font("tiny").render(key_labels.get(btn["state"], ""), True, UI_SUBTEXT)
        s.blit(kl, (icon_cx - kl.get_width() // 2, rect.y + 214))


# ── Icon drawing helpers ───────────────────────────────────────────────────────

def _draw_book_icon(s, cx, cy, hovered):
    col = ASTAR_MAN_COL if hovered else (160, 140, 100)
    # Left page
    pygame.draw.rect(s, col, (cx - 38, cy - 30, 36, 50), border_radius=3)
    # Right page
    pygame.draw.rect(s, col, (cx + 2,  cy - 30, 36, 50), border_radius=3)
    # Spine
    pygame.draw.rect(s, UI_BG, (cx - 3, cy - 30, 6, 50))
    pygame.draw.rect(s, (*col, 150), (cx - 3, cy - 30, 6, 50))
    # Page lines
    page_line = (UI_BG[0], UI_BG[1], UI_BG[2])
    for dy in (0, 10, 20):
        pygame.draw.line(s, page_line, (cx - 34, cy - 18 + dy), (cx - 6, cy - 18 + dy), 2)
        pygame.draw.line(s, page_line, (cx + 6,  cy - 18 + dy), (cx + 34, cy - 18 + dy), 2)
    # A* star label on right page
    star = font("small").render("A*", True, UI_BG)
    s.blit(star, (cx + 14, cy + 4))


def _draw_maze_icon(s, cx, cy, hovered):
    """Draw a tiny 5×5 maze grid with a winding highlighted path."""
    col_wall  = ASTAR_MAN_COL if hovered else (100, 160, 100)
    col_path  = UI_SUCCESS     if hovered else (60, 140, 60)
    cell = 10
    ox   = cx - 2 * cell - cell // 2
    oy   = cy - 2 * cell - cell // 2
    # Draw grid outline
    pygame.draw.rect(s, col_wall, (ox - 1, oy - 1, 5 * cell + 2, 5 * cell + 2), 2)
    # Internal walls (hand-crafted to look maze-like)
    walls = [
        # horizontal walls (block going down)
        ((0,0),(1,0)), ((2,0),(3,0)),
        ((1,1),(2,1)), ((3,1),(4,1)),
        ((0,2),(0,3)), ((2,2),(2,3)),((4,2),(4,3)),
        ((1,3),(2,3)), ((3,3),(3,4)),
    ]
    for (c1,r1),(c2,r2) in walls:
        x1 = ox + c1 * cell; y1 = oy + r1 * cell
        x2 = ox + c2 * cell; y2 = oy + r2 * cell
        pygame.draw.line(s, col_wall, (x1, y1), (x2, y2), 2)
    # Path (start bottom-left → end top-right)
    path_cells = [(0,4),(0,3),(1,3),(1,2),(2,2),(2,1),(3,1),(3,0),(4,0)]
    for c, r in path_cells:
        pygame.draw.rect(s, col_path,
                         (ox + c*cell+2, oy + r*cell+2, cell-4, cell-4), border_radius=2)
    # Start / end dots
    pygame.draw.circle(s, START_COLOR, (ox + 0*cell + cell//2, oy + 4*cell + cell//2), 4)
    pygame.draw.circle(s, END_COLOR,   (ox + 4*cell + cell//2, oy + 0*cell + cell//2), 4)


def _draw_custom_icon(s, cx, cy, hovered):
    col = ASTAR_EUC_COL if hovered else (120, 120, 160)
    # Pencil body
    body_pts = [
        (cx - 8, cy - 34), (cx + 8, cy - 34),
        (cx + 8, cy + 24), (cx - 8, cy + 24),
    ]
    pygame.draw.polygon(s, col, body_pts)
    # Pencil tip
    tip_pts = [(cx - 8, cy + 24), (cx + 8, cy + 24), (cx, cy + 40)]
    pygame.draw.polygon(s, (220, 180, 100), tip_pts)
    # Eraser
    pygame.draw.rect(s, (220, 100, 100), (cx - 8, cy - 44, 16, 10), border_radius=2)
    # Lines on pencil body
    for dy in (-20, -8, 4):
        pygame.draw.line(s, (*UI_BG, 180), (cx - 6, cy + dy), (cx + 6, cy + dy), 1)
    # Gear/cog to the right of pencil
    gcx, gcy, gr = cx + 28, cy - 10, 14
    pygame.draw.circle(s, col, (gcx, gcy), gr, 2)
    pygame.draw.circle(s, col, (gcx, gcy), 5)
    for a in range(0, 360, 45):
        rad = math.radians(a)
        gx1 = gcx + int((gr - 2) * math.cos(rad))
        gy1 = gcy + int((gr - 2) * math.sin(rad))
        gx2 = gcx + int((gr + 4) * math.cos(rad))
        gy2 = gcy + int((gr + 4) * math.sin(rad))
        pygame.draw.line(s, col, (gx1, gy1), (gx2, gy2), 3)


# Bring colours into scope for icon drawing
from src.config import START_COLOR, END_COLOR  # noqa: E402

File: web/src/states/menuState.js
/**
 * menuState.js — Main menu: animated backdrop, 3 icon buttons, algorithm badges.
 * Mirrors src/states/menu_state.py
 */

import { SCREEN_WIDTH, SCREEN_HEIGHT, C, drawRoundRect, drawText } from '../renderer.js';

// ── Button layout ─────────────────────────────────────────────────────────────
const BTN_W    = 220;
const BTN_H    = 230;
const BTN_GAP  = 44;
const TOTAL_W  = 3 * BTN_W + 2 * BTN_GAP;
const BTN_X0   = Math.floor((SCREEN_WIDTH - TOTAL_W) / 2);
// Push cards lower so header breathes, cards sit in comfortable lower half
const BTN_TOP  = 210;

const BTNS = [
    { label: 'Tutorial', sub: 'Learn A* step by step',  target: 'TUTORIAL', hint: 'T' },
    { label: 'Maze',     sub: 'Race through 4 layers',  target: 'MAZE_L1',  hint: '1' },
    { label: 'Custom',   sub: 'Free-form visualizer',   target: 'CUSTOM',   hint: 'C' },
];

// Layers tracked for completion badges on the Maze card
const LAYERS = ['MAZE_L1', 'MAZE_L2', 'MAZE_L3', 'MAZE_L4'];

const KEY_MAP = { t: 'TUTORIAL', '1': 'MAZE_L1', c: 'CUSTOM' };

// ── Factory ───────────────────────────────────────────────────────────────────

export function createMenuState(ctx, goTo, gameData) {
    let anim  = 0;
    let mx = 0, my = 0;

    const rects = BTNS.map((_, i) => ({
        x: BTN_X0 + i * (BTN_W + BTN_GAP), y: BTN_TOP, w: BTN_W, h: BTN_H,
    }));

    function enter() { anim = 0; }
    function update(dt) { anim += dt; }
    function handleMouseMove(x, y) { mx = x; my = y; }

    function handleClick(x, y) {
        for (let i = 0; i < BTNS.length; i++)
            if (_hit(rects[i], x, y)) { goTo(BTNS[i].target); return; }
    }

    function handleKey(key) {
        const target = KEY_MAP[key.toLowerCase()];
        if (target) goTo(target);
    }

    // ── Draw ──────────────────────────────────────────────────────────────────

    function draw() {
        // Light academic background
        ctx.fillStyle = '#f4f6fc';
        ctx.fillRect(0, 0, SCREEN_WIDTH, SCREEN_HEIGHT);

        // Subtle grid pattern (academic graph-paper feel)
        ctx.strokeStyle = 'rgba(160,170,210,0.25)';
        ctx.lineWidth   = 1;
        for (let gx = 0; gx < SCREEN_WIDTH; gx += 40) {
            ctx.beginPath(); ctx.moveTo(gx, 0); ctx.lineTo(gx, SCREEN_HEIGHT); ctx.stroke();
        }
        for (let gy = 0; gy < SCREEN_HEIGHT; gy += 40) {
            ctx.beginPath(); ctx.moveTo(0, gy); ctx.lineTo(SCREEN_WIDTH, gy); ctx.stroke();
        }

        // Animated accent dots at grid intersections
        for (let gx = 0; gx < SCREEN_WIDTH; gx += 40) {
            for (let gy = 0; gy < SCREEN_HEIGHT; gy += 40) {
                const pulse = 0.12 + 0.06 * Math.sin(anim * 0.6 + gx * 0.04 + gy * 0.04);
                ctx.fillStyle = `rgba(37,99,235,${pulse.toFixed(3)})`;
                ctx.beginPath(); ctx.arc(gx, gy, 1.5, 0, Math.PI * 2); ctx.fill();
            }
        }

        const cx = SCREEN_WIDTH / 2;
        ctx.textBaseline = 'top';
        ctx.textAlign    = 'center';

        // ── Header bar (academic paper style) ────────────────────────────────
        ctx.fillStyle = '#6b7d96';
        ctx.fillRect(0, 0, SCREEN_WIDTH, 36);
        ctx.font      = '500 13px "Inter", Arial, sans-serif';
        ctx.fillStyle = 'rgba(255,255,255,0.7)';
        ctx.textAlign = 'left'; ctx.textBaseline = 'middle';
        ctx.fillText('CS5800 · Algorithms & Data Structures', 20, 18);
        ctx.textAlign = 'right';
        ctx.fillText('Interactive Learning Tool', SCREEN_WIDTH - 20, 18);

        // ── Title block ───────────────────────────────────────────────────────
        ctx.textAlign = 'center'; ctx.textBaseline = 'top';

        // English title (enlarged, moved up since CJK removed)
        ctx.font      = '700 38px "Inter", Arial, sans-serif';
        ctx.fillStyle = '#3a4a6a';
        ctx.fillText('The Path of Algorithms', cx, 54);

        // Subtitle
        ctx.font      = '400 14px "Inter", Arial, sans-serif';
        ctx.fillStyle = C.uiSubtext;
        ctx.fillText('A*  vs  Dijkstra  ·  Interactive Maze Learning', cx, 102);

        // ── Algorithm badge pills ─────────────────────────────────────────────
        _drawBadges(cx);

        // ── Three main cards ──────────────────────────────────────────────────
        for (let i = 0; i < BTNS.length; i++) _drawButton(i);

        // ── Key-shortcut strip ────────────────────────────────────────────────
        const hintY = BTN_TOP + BTN_H + 18;
        ctx.font      = '400 13px "Inter", Arial, sans-serif';
        ctx.fillStyle = C.uiSubtext;
        ctx.textAlign = 'center';
        ctx.fillText(
            '[T] Tutorial  ·  [1] Maze  ·  [C] Custom',
            cx, hintY,
        );

        // ── Unlocked tomes strip (bottom-left) ────────────────────────────────
        // Bottom border
        ctx.fillStyle = '#d0d4e8';
        ctx.fillRect(0, SCREEN_HEIGHT - 34, SCREEN_WIDTH, 1);
        ctx.fillStyle = '#f8f9fd';
        ctx.fillRect(0, SCREEN_HEIGHT - 33, SCREEN_WIDTH, 33);

        if (gameData.tomes?.length) {
            ctx.textAlign    = 'left';
            ctx.textBaseline = 'middle';
            ctx.font         = '600 11px "Inter", Arial, sans-serif';
            ctx.fillStyle    = C.uiSubtext;
            ctx.fillText('UNLOCKED', 20, SCREEN_HEIGHT - 17);
            let tx = 20 + ctx.measureText('UNLOCKED').width + 12;
            for (const t of gameData.tomes) {
                const col = t === 'manhattan' ? C.astarMan
                          : t === 'euclidean' ? C.astarEuc
                          : C.astarHex;
                const label = t.charAt(0).toUpperCase() + t.slice(1);
                ctx.font = '500 11px "Inter", Arial, sans-serif';
                const lw = ctx.measureText(label).width;
                ctx.fillStyle   = col + '20';
                ctx.strokeStyle = col + '80';
                ctx.lineWidth   = 1;
                ctx.beginPath(); ctx.roundRect(tx, SCREEN_HEIGHT - 26, lw + 16, 18, 4);
                ctx.fill(); ctx.stroke();
                ctx.fillStyle = col;
                ctx.fillText(label, tx + 8, SCREEN_HEIGHT - 17);
                tx += lw + 28;
            }
        } else {
            ctx.textAlign = 'left'; ctx.textBaseline = 'middle';
            ctx.font = '400 12px "Inter", Arial, sans-serif';
            ctx.fillStyle = C.uiSubtext;
            ctx.fillText('Complete Tutorial to unlock algorithm tomes.', 20, SCREEN_HEIGHT - 17);
        }
    }

    function _drawBadges(cx) {
        const badges = [
            { text: 'Dijkstra  h=0',              color: C.dijkstra },
            { text: 'A* Manhattan  h=|Δx|+|Δy|',  color: C.astarMan },
            { text: 'A* Euclidean  h=√(Δx²+Δy²)', color: C.astarEuc },
        ];
        ctx.font = '500 12px "Inter", Arial, sans-serif';
        const gap = 12, padX = 10, badgeY = 150, badgeH = 24;
        const widths = badges.map(b => ctx.measureText(b.text).width + padX * 2);
        const total  = widths.reduce((a, w) => a + w + gap, -gap);
        let bx = cx - total / 2;
        for (let i = 0; i < badges.length; i++) {
            const { text, color } = badges[i];
            const bw = widths[i];
            ctx.fillStyle   = color + '18';
            ctx.strokeStyle = color + '80';
            ctx.lineWidth   = 1;
            ctx.beginPath(); ctx.roundRect(bx, badgeY, bw, badgeH, 5);
            ctx.fill(); ctx.stroke();
            ctx.fillStyle    = color;
            ctx.textAlign    = 'left';
            ctx.textBaseline = 'middle';
            ctx.fillText(text, bx + padX, badgeY + badgeH / 2);
            bx += bw + gap;
        }
    }

    function _drawButton(idx) {
        const { x, y, w, h } = rects[idx];
        const btn   = BTNS[idx];
        const hov   = _hit(rects[idx], mx, my);
        const pulse = 0.5 + 0.5 * Math.sin(anim * 1.8 + idx * 1.1);
        const icx   = x + w / 2;
        const icy   = y + 72;

        // Card — white surface with drop shadow illusion via border
        ctx.fillStyle = hov ? '#f0f4ff' : '#ffffff';
        ctx.beginPath(); ctx.roundRect(x, y, w, h, 12); ctx.fill();

        // Card border — accent on hover, subtle otherwise
        const accentCol = idx === 0 ? C.astarMan : idx === 1 ? C.dijkstra : C.astarEuc;
        if (hov) {
            ctx.strokeStyle = accentCol;
            ctx.lineWidth   = 2;
        } else {
            ctx.strokeStyle = C.uiBorder;
            ctx.lineWidth   = 1;
        }
        ctx.beginPath(); ctx.roundRect(x, y, w, h, 12); ctx.stroke();

        // Accent top bar (always visible, brightens on hover)
        ctx.fillStyle = accentCol + (hov ? 'ee' : '99');
        ctx.beginPath(); ctx.roundRect(x, y, w, 4, [12, 12, 0, 0]); ctx.fill();

        // Icon
        if      (idx === 0) _drawBookIcon(icx, icy, hov);
        else if (idx === 1) _drawMazeIcon(icx, icy, hov);
        else                _drawCustomIcon(icx, icy, hov);

        // Label
        ctx.font         = '700 18px "Inter", Arial, sans-serif';
        ctx.fillStyle    = hov ? C.uiHeading : C.uiText;
        ctx.textAlign    = 'center';
        ctx.textBaseline = 'top';
        ctx.fillText(btn.label, icx, y + 146);

        // Sub-label
        ctx.font      = '400 13px "Inter", Arial, sans-serif';
        ctx.fillStyle = hov ? C.uiHilight : C.uiSubtext;
        ctx.fillText(btn.sub, icx, y + 170);

        // Completion dots for Maze card (layers L1-L4)
        if (idx === 1) {
            const dotGap = 16, dotR = 4;
            const dotsW  = LAYERS.length * dotGap - (dotGap - dotR * 2);
            let dx = icx - dotsW / 2 + dotR;
            const dotY  = y + 196;
            for (const lvl of LAYERS) {
                const done = gameData.completedLevels?.includes(lvl);
                ctx.fillStyle = done ? C.uiSuccess : C.uiBorder;
                ctx.beginPath(); ctx.arc(dx, dotY, dotR, 0, Math.PI * 2); ctx.fill();
                if (done) {
                    ctx.fillStyle = 'rgba(255,255,255,0.9)';
                    ctx.beginPath(); ctx.arc(dx - 1, dotY - 1, 1.5, 0, Math.PI * 2); ctx.fill();
                }
                dx += dotGap;
            }
        }

        // Key hint pill (keyboard shortcut badge)
        const kx = x + w - 30, ky = y + h - 28;
        ctx.fillStyle   = '#e8ecf8';
        ctx.strokeStyle = C.uiBorder;
        ctx.lineWidth   = 1;
        ctx.beginPath(); ctx.roundRect(kx, ky, 22, 18, 4); ctx.fill(); ctx.stroke();
        ctx.font      = '600 11px "Inter", Arial, sans-serif';
        ctx.fillStyle = C.uiSubtext;
        ctx.textAlign = 'center'; ctx.textBaseline = 'middle';
        ctx.fillText(btn.hint, kx + 11, ky + 9);
    }

    // ── Icon drawing ──────────────────────────────────────────────────────────

    function _drawBookIcon(cx, cy, hov) {
        const col     = hov ? C.astarMan : 'rgb(160,140,100)';
        const spineAlpha = hov ? 'rgba(255,220,30,0.6)' : 'rgba(160,140,100,0.6)';

        ctx.fillStyle = col;
        ctx.beginPath(); ctx.roundRect(cx - 38, cy - 30, 36, 50, 3); ctx.fill();
        ctx.beginPath(); ctx.roundRect(cx +  2, cy - 30, 36, 50, 3); ctx.fill();

        // Spine
        ctx.fillStyle = C.uiBg;       ctx.fillRect(cx - 3, cy - 30, 6, 50);
        ctx.fillStyle = spineAlpha;   ctx.fillRect(cx - 3, cy - 30, 6, 50);

        // Page lines
        ctx.strokeStyle = C.uiBg; ctx.lineWidth = 2;
        for (const dy of [0, 10, 20]) {
            _stroke(cx - 34, cy - 18 + dy, cx - 6, cy - 18 + dy);
            _stroke(cx +  6, cy - 18 + dy, cx + 34, cy - 18 + dy);
        }

        // "A*" label
        ctx.font         = 'bold 13px Arial';
        ctx.fillStyle    = C.uiBg;
        ctx.textAlign    = 'left';
        ctx.textBaseline = 'top';
        ctx.fillText('A*', cx + 14, cy + 4);
    }

    function _drawMazeIcon(cx, cy, hov) {
        const colWall = hov ? C.astarMan : 'rgb(100,160,100)';
        const colPath = hov ? C.uiSuccess : 'rgb(60,140,60)';
        const cell = 10;
        const ox = cx - 2.5 * cell, oy = cy - 2.5 * cell;

        // Outline
        ctx.strokeStyle = colWall; ctx.lineWidth = 2;
        ctx.beginPath(); ctx.rect(ox - 1, oy - 1, 5*cell+2, 5*cell+2); ctx.stroke();

        // Walls
        for (const [c1,r1,c2,r2] of [
            [0,0,1,0],[2,0,3,0], [1,1,2,1],[3,1,4,1],
            [0,2,0,3],[2,2,2,3],[4,2,4,3], [1,3,2,3],[3,3,3,4],
        ]) { _stroke(ox+c1*cell, oy+r1*cell, ox+c2*cell, oy+r2*cell); }

        // Path cells
        ctx.fillStyle = colPath;
        for (const [c,r] of [[0,4],[0,3],[1,3],[1,2],[2,2],[2,1],[3,1],[3,0],[4,0]]) {
            ctx.beginPath(); ctx.roundRect(ox+c*cell+2, oy+r*cell+2, cell-4, cell-4, 2); ctx.fill();
        }

        // Start / End dots
        ctx.fillStyle = C.start;
        ctx.beginPath(); ctx.arc(ox+cell/2, oy+4*cell+cell/2, 4, 0, Math.PI*2); ctx.fill();
        ctx.fillStyle = C.end;
        ctx.beginPath(); ctx.arc(ox+4*cell+cell/2, oy+cell/2, 4, 0, Math.PI*2); ctx.fill();
    }

    function _drawCustomIcon(cx, cy, hov) {
        const col = hov ? C.astarEuc : 'rgb(120,120,160)';

        // Pencil body
        ctx.fillStyle = col;
        ctx.beginPath();
        ctx.moveTo(cx-8, cy-34); ctx.lineTo(cx+8, cy-34);
        ctx.lineTo(cx+8, cy+24); ctx.lineTo(cx-8, cy+24);
        ctx.closePath(); ctx.fill();

        // Tip
        ctx.fillStyle = 'rgb(220,180,100)';
        ctx.beginPath();
        ctx.moveTo(cx-8, cy+24); ctx.lineTo(cx+8, cy+24); ctx.lineTo(cx, cy+40);
        ctx.closePath(); ctx.fill();

        // Eraser
        ctx.fillStyle = 'rgb(220,100,100)';
        ctx.beginPath(); ctx.roundRect(cx-8, cy-44, 16, 10, 2); ctx.fill();

        // Body lines
        ctx.strokeStyle = 'rgba(18,18,28,0.7)'; ctx.lineWidth = 1;
        for (const dy of [-20, -8, 4]) _stroke(cx-6, cy+dy, cx+6, cy+dy);

        // Gear
        const [gcx, gcy, gr] = [cx + 28, cy - 10, 14];
        ctx.strokeStyle = col; ctx.lineWidth = 2;
        ctx.beginPath(); ctx.arc(gcx, gcy, gr, 0, Math.PI*2); ctx.stroke();
        ctx.fillStyle = col;
        ctx.beginPath(); ctx.arc(gcx, gcy, 5, 0, Math.PI*2); ctx.fill();
        ctx.lineWidth = 3;
        for (let a = 0; a < 360; a += 45) {
            const r = (Math.PI / 180) * a;
            _stroke(gcx+(gr-2)*Math.cos(r), gcy+(gr-2)*Math.sin(r),
                    gcx+(gr+4)*Math.cos(r), gcy+(gr+4)*Math.sin(r));
        }
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    function _stroke(x1, y1, x2, y2) {
        ctx.beginPath(); ctx.moveTo(x1, y1); ctx.lineTo(x2, y2); ctx.stroke();
    }

    return { enter, update, draw, handleKey, handleClick, handleMouseMove };
}

function _hit(r, x, y) {
    return x >= r.x && x <= r.x + r.w && y >= r.y && y <= r.y + r.h;
}
